HerPublicWebsite/Services/QuestionnaireService.cs

```diff
@@ -169,16 +168,20 @@ public async Task<Questionnaire> RecordNotificationConsentAsync(bool consentGran
             notificationEmailAddress,
             confirmationConsentGranted,
             confirmationEmailAddress);
-        SaveQuestionnaireToSession(questionnaire);
+        await SaveQuestionnaireToSession(questionnaire);
         return questionnaire;
     }
 
-    public void SaveQuestionnaireToSession(Questionnaire questionnaire)
+    public async Task SaveQuestionnaireToSession(Questionnaire questionnaire)
     {
         var hasNotSavedQuestionnaireBefore =
             httpContextAccessor.HttpContext!.Session.GetString(SessionKeyQuestionnaire) == null;
-        if (hasNotSavedQuestionnaireBefore) sessionRecorderService.RecordNewSessionStarted();
-
+        if (hasNotSavedQuestionnaireBefore)
+        {
+            var newSessionStarted = await sessionRecorderService.RecordNewSessionStarted();
+            questionnaire.SessionId = newSessionStarted.Id;

```

Comment:
should this use QuestionnaireUpdater?

Reply:
Should we be saving the questionnaire here?

Reply:
Can confirm, I've looked into it and we're saving the questionnaire to the server-side session over the next two lines.

Reply:
I mean to the database rather than the session - we're updating the `SessionId` property here but not persisting it as far as I can see?

Reply:
If you mean the ID in the Session table:

- That is the source of the assignment here (`newSessionStarted.Id` is returned when the entry in the Sessions table is created and saved, and the ID is auto-generated there)

If you mean the Session ID saved to the questionnaire:

- We don't save the questionnaire to the database

Reply:
After discussing: 

Summary:
Cookie Session ID -> References -> HTTPContext.Session -> Contains -> Questionnaire -> Contains -> Session ID -> references -> Database Session Table ID

HTTPContext.Session contains the questionnaire and means it persists. The ID from the database is stored in the questionnaire, which is stored in the HTTPContext.Session, meaning the ID persists.

Full explanation:
The Questionnaire is not stored in the database. During the journey, the questionnaire is stored in HTTPContext.Session as a serialized string. This is updated after each question. In the code above, the Session ID from the Sessions table in the database is saved to the HTTPContext.Session string (which causes it to persist). Only at the end of the journey is the questionnaire (somewhat) saved to the database in the form of a Referral Request.

This ID is used to reference the "database session" using the "HTTPContext session", allowing us to edit the Session's DB entry later, when the journey completes.

Worth noting, there is also a "Session ID" which is stored client side in a cookie, and on the server. This is separate from the ID stored in the questionnaire and the database. This is used to link the client's browser to the HTTPContext.Session.